### PR TITLE
Gemfile cleanup & Whirlpool encoder/gem removal

### PR DESCRIPTION
**Pull request by tolly765**
Hello!

As discussed previously, this PR is to clean up the Gemfile for SecGen and allow it to install without any additional modifications needed. This required a small bundle of encoders to have their dependencies updated, and also unfortunately required the removal of the "digest-whirlpool" gem along with the corresponding encoder.

See https://github.com/cliffe/SecGen/commit/ea8e19c59769dea05dcf8d73db7e7f09aad71800 for the updated encoders affected by this Gemfile change. Some test scenarios using these encoders may need to be run to ensure that these encoders still function as intended, but the testing done on this side appears to show that the encoders work as intended.

Any changes are more than welcome, and hopefully this will help

### CODE REVIEW

**Comment by thomashaw** [comment]
These encoders are used in quite a lot of spots across the course, so we're
going to need to reimplement with an alternative gem (or roll our own lib,
interfacing with a python encoder running locally perhaps?). It'd be nice
to have that in place before pulling this into master, just in case the
regression doesn't get resolved before the next time we need the hash
functions.

More than happy for alternative suggestions.

Cheers,
Tom

**Comment by aduyinuo** [comment]
Hi!

Does this version has the rescue standardError issue? Which ruby version and ubuntu version are required?

Thanks!

File: modules/encoders/hash/sha1/secgen_local/local.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/ruby
 require_relative '../../../../../lib/objects/local_hash_encoder.rb'
+require 'openssl'
```
Comment (cliffe) [inline modules/encoders/hash/sha1/secgen_local/local.rb:3]: Why does this change sometimes use `require 'openssl'` and sometimes `require 'sha3'`?

The Gemfile has `sha3` added

File: Gemfile [hunk 2]
```diff
@@ -33,16 +33,16 @@ gem 'huffman', :git => "https://github.com/fgosew/huffman"
 gem 'ruby-graphviz'
 gem 'rsa'
 gem 'gpgmeh'
-gem 'digest-sha3', :git => "http://github.com/izetex/digest-sha3-ruby"
+gem 'sha3'
 gem 'packetfu'
 gem 'net-ntp'
 gem 'CFPropertyList'
 gem 'artii'
+gem 'sqlite3'
```
Comment (cliffe) [inline Gemfile:41]: please remove this from the PR until we have some code in SecGen that actually uses this

**Review by cliffe — CHANGES_REQUESTED** [review]

**Comment by cliffe** [comment]
Thanks, @tolly765, could you please take a look at the above comments.

If we only loose Whirlpool hashes, and use a different library for SHA3, that shouldn't break any existing scenarios. If thatmakes it easier to install SecGen, we can go ahead and look to add additional hash functions as a separate step.

@aduyinuo I've not personally seen that error, but this reportedly makes installing gems easier to resolve.

File: modules/encoders/hash/sha1/secgen_local/local.rb (hunk 1 above)
Comment (tolly765) [reply modules/encoders/hash/sha1/secgen_local/local.rb:3]: Digest-SHA3 (the previous library) was one of the few gems causing problems with the install process - most likely due to the fact that it was [last updated in 2015](https://rubygems.org/gems/digest-sha3/versions/1.1.0).The OpenSSL library has some support for SHA3, but it does not cover all the variations present within SecGen - hence why a replacement SHA3 gemfile was sourced.
For reference, here is a list of [all the Digests available in OpenSSL](https://ruby.github.io/openssl/OpenSSL/Digest.html).

File: Gemfile (hunk 2 above)
Comment (tolly765) [reply Gemfile:41]: My apologies, I thought I got rid of that before pushing the PR through. I'll remove this.

**Comment by tolly765** [comment]
The requested changes have been made, and explanations have been written into the comments above. If there is anything else that needs sorting before merge, please let me know 👍

**Comment by cliffe** [comment]
Thanks, @tolly765. I think you can leave all the openssl changes off, since SHA1 and 2 are provided by stdlib. It's just the SHA3 calls you need to change?

**Comment by tolly765** [comment]
See https://github.com/cliffe/SecGen/commit/1205a48397bbfbf60778ea40431301559e77c127 for the reversal of the OpenSSL changes. Please note that doing this though required a separate change to local_encoding_functions to stop the following error from occurring:

```
SHA1 Encoder
Reading args from STDIN
/home/tolly765/Github/SecGen/lib/objects/local_encoding_functions.rb:28:in `force_encoding': can't modify frozen String: "test" (FrozenError)
        from /home/tolly765/Github/SecGen/lib/objects/local_encoding_functions.rb:28:in `block in array_to_utf8'
        from /home/tolly765/Github/SecGen/lib/objects/local_encoding_functions.rb:26:in `map'
        from /home/tolly765/Github/SecGen/lib/objects/local_encoding_functions.rb:26:in `array_to_utf8'
        from /home/tolly765/Github/SecGen/lib/objects/local_string_encoder.rb:144:in `block in enforce_utf8'
        from /home/tolly765/Github/SecGen/lib/objects/local_string_encoder.rb:141:in `each'
        from /home/tolly765/Github/SecGen/lib/objects/local_string_encoder.rb:141:in `enforce_utf8'
        from /home/tolly765/Github/SecGen/lib/objects/local_string_encoder.rb:120:in `run'
        from /home/tolly765/Github/SecGen/modules/encoders/hash/sha1/secgen_local/local.rb:15:in `<main>'
```

**Comment by cliffe** [comment]
Great!

**Pull request merged**